### DISCUSSION TITLE
What CF cards are supported?

### DISCUSSION THREAD

**Discussion by sparkybrit** [wwarthen/RomWBW#292] [Q&A] [answered]
I'm using MK4_rom.bin on a Mark IV.
I've tried a few CF cards: some work, some don't.
Some UDMA 6 cards work and some don't. 
UDMA 7 cards don't work.

Does the BIOS use PIO or UDMA? 

PIO
Mode 6 - 25MB/s
Mode 5 - 20MB/s
Mode 4 - 16.7MB/s
Mode 3 – 11.1MB/s
Mode 2 – 8.3MB/s
Mode 1 – 5.2MB/s

UDMA
Mode 7 - 166MB/s
Mode 6 - 133MB/s
Mode 5 - 100MB/s
Mode 4 - 66.7 MB/s
Mode 3 - 44.4MB/s
Mode 2 - 33.3MB/s
Mode 1 - 25MB/s

What cards are supposed to work?

**Comment by wwarthen** [comment]
All cards should work.  There are no known CF Card software incompatibilities.

RomWBW uses PIO to access CF cards.  All cards should be PIO compatible (I have not encountered any that are not).

The typical issue with compatibility has to do with the CF Card interface.  Newer/larger cards tend to have very fast signal transitions that are not always handled well by the hardware interface.  The standard RC2014 CF Card module is particularly problematic because it has no signal buffering.

What platform are you using?  What version of RomWBW are you using?

If you are comfortable building a custom ROM, you can add the following line to your configuration file to enable detailed debug output for CF Cards:

`IDETRACE      .SET        3`

Thanks,

Wayne

**Reply by sparkybrit** [reply]
I'm using Mk4_rom V3.0.1 on a Mark IV

**Reply by wwarthen** [reply]
> I'm using Mk4_rom V3.0.1 on a Mark IV

OK.  I have had pretty good luck with that hardware interface.  It does require that the CF Card support 8-bit I/O, but they are all supposed to do so.

Do you want me to send you a pre-built ROM image with the debug output enabled?

Thanks,

Wayne

**Reply by sparkybrit** [reply]
No, I should be able to build that myself :-)

**Reply by wwarthen** [reply]
> No, I should be able to build that myself :-)

Great.  I should be able to tell from the DEBUG output at startup what is happening with specific cards.   If it is hardware, you will typically see that the card is unresponsive to attempts to wake up and initialize the card.  If there is a software protocol issue, you will see some packet exchanges followed by something going wrong.

Feel free to forward any debug output.

Thanks,

Wayne

**Reply by sparkybrit** [reply]
I checked out tag/v3.0.1, ran the build on Mac OS. The Build completed with no errors but the rom is not the same as the rom in the binary distro and it doesn't boot.
I could try to debug the MacOS build but might be easier in the short term if you send me a rom with idetrace turned on. 
Still up for that?

**Reply by wwarthen** [reply]
No problem.  I have attached a zip file that includes the ROM image (.rom) with IDE debugging enabled.  I tried it on my Mark IV system and it worked as expected (got debug output).

The zip file also includes a .com version of the build.  If you have a way to transfer this to your system, you can run the .com file under CP/M and it will load the debug ROM on-the-fly without burning a new ROM.  There are scenarios where this mechanism will fail, so don't worry if that happens.  Just revert to programming the ROM.

Note that the provided build is based on the current dev branch.  This branch is quite stable and is probably best for testing the CF Cards.

Let me know how you do.

Thanks,

Wayne

[MK4_dbg.zip](https://github.com/wwarthen/RomWBW/files/9033465/MK4_dbg.zip)

**Comment by sparkybrit** [comment, marked as answer]
Thanks for the debug rom. 

FYI This is what the (working) 8GB SanDisk Extreme 60MB/s UDMA 6 CF card looks like with IDETRACE set:

RomWBW HBIOS v3.1.1-pre.177, 2022-07-02

MARK IV (ide dbg) Z8S180-N @ 18.432MHz IO=0x40
0 MEM W/S, 2 I/O W/S, INT MODE 2, Z180 MMU
512KB ROM, 512KB RAM
ROM VERIFY: 00 00 00 00 PASS

ASCI0: IO=0x40 ASCI W/BRG MODE=38400,8,N,1
ASCI1: IO=0x41 ASCI W/BRG MODE=38400,8,N,1
DSRTC: MODE=STD IO=0x8A Sun 2022-07-03 09:32:28 CHARGE=ON
CVDU: MODE=ECB EGA IO=0xE4 NOT PRESENT
VGA: IO=0xE2 NOT PRESENT
MD: UNITS=2 ROMDISK=384KB RAMDISK=256KB
FD: MODE=DIDE IO=0x2A NOT PRESENT
IDE: IO=0x80 MODE=MK4
IDE0: RESET FAKE HARD CONFIG
IDE0: SELUNIT
IDE0: PROBE E0 50 [50 E0 00 00 01 01 01] [50 E0 00 00 01 01 01] 50 01 01 00 00
IDE0: SETFEAT E0 01 EF --> 50
IDE0: IDDEV E0 EC --> 58 GETBUF 50
604D 8A 84 A5 3C 00 00 10 00 00 00 40 02 3F 00 EE 00  ..<......@.?..
605D B0 C9 00 00 20 20 20 20 43 20 5A 50 32 30 36 32  ..    C ZP2062
606D 30 31 30 32 35 35 37 31 02 00 02 00 04 00 44 48  01025571......DH
607D 20 58 2E 35 38 30 61 53 44 6E 73 69 20 6B 44 53   X.580aSDnsi kDS
608D 46 43 2D 58 30 30 47 38 20 20 20 20 20 20 20 20  FC-X00G8        
609D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 01 80                ..
60AD 00 00 00 23 00 00 00 02 00 00 07 00 A5 3C 10 00  ...#........<..
60BD 3F 00 B0 C9 EE 00 00 01 B0 C9 EE 00 00 00 07 00  ?.........
60CD 03 00 78 00 78 00 78 00 78 00 00 00 00 00 00 00  ..x.x.x.x.......
60DD 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
60ED 10 00 00 00 20 70 0C 74 00 40 00 70 04 04 00 40  .... p.t.@.p...@
60FD 3F 00 00 00 00 00 00 00 FE FF 00 00 00 00 00 00  ?.............
610D 00 00 00 00 00 00 00 00 B0 C9 EE 00 00 00 00 00  .............
611D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
612D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
613D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
614D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
615D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
616D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
617D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
618D F4 81 00 00 00 00 82 00 00 00 00 00 00 00 00 50  ..............P
619D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
61AD 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
61BD 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
61CD 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
61DD 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
61ED 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
61FD 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
620D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
621D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
622D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
623D 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00  ................
IDE0: SELUNIT
IDE0: PROBE F0 00 [00 F0 00 00 01 01 00] [00 F0 00 00 01 01 00] 00
IDE0: NO MEDIA [00 F0 00 00 01 01 00]
IDE0: 8-BIT LBA BLOCKS=0x00EEC9B0 SIZE=7641MB
IDE1: NO MEDIA
SD: MODE=MK4 FAST OPR=0x89 CNTR=0x4A TRDR=0x4B DEVICES=1
SD0: SDHC NAME=SS32G BLOCKS=0x03B72400 SIZE=30436MB
PRP: IO=0xA8 NOT PRESENT

Unit        Device      Type              Capacity/Mode
----------  ----------  ----------------  --------------------
Char 0      ASCI0:      RS-232            38400,8,N,1
Char 1      ASCI1:      RS-232            38400,8,N,1
Disk 0      MD0:        RAM Disk          256KB,LBA
Disk 1      MD1:        ROM Disk          384KB,LBA
Disk 2      IDE0:       CompactFlash      7641MB,LBA
Disk 3      IDE1:       Hard Disk         --
Disk 4      SD0:        SD Card           30436MB,LBA


MARK IV (ide dbg) Boot Loader

Boot [H=Help]: h
---------------------------------------------
This is what the (failing) 8GB Transcend CF x133 card looks like with TRACE set:-

RomWBW HBIOS v3.1.1-pre.177, 2022-07-02

MARK IV (ide dbg) Z8S180-N @ 18.432MHz IO=0x40
0 MEM W/S, 2 I/O W/S, INT MODE 2, Z180 MMU
512KB ROM, 512KB RAM
ROM VERIFY: 00 00 00 00 PASS

ASCI0: IO=0x40 ASCI W/BRG MODE=38400,8,N,1
ASCI1: IO=0x41 ASCI W/BRG MODE=38400,8,N,1
DSRTC: MODE=STD IO=0x8A Sun 2022-07-03 09:36:01 CHARGE=ON
CVDU: MODE=ECB EGA IO=0xE4 NOT PRESENT
VGA: IO=0xE2 NOT PRESENT
MD: UNITS=2 ROMDISK=384KB RAMDISK=256KB
FD: MODE=DIDE IO=0x2A NOT PRESENT
IDE: IO=0x80 MODE=MK4
IDE0: RESET FAKE HARD CONFIG
IDE0: SELUNIT
IDE0: PROBE E0 00 [00 00 00 00 00 00 00] [00 00 00 00 00 00 00] 00
IDE0: NO MEDIA [00 00 00 00 00 00 00]
IDE0: SELUNIT
IDE0: PROBE F0 78
IDE0: NO MEDIA [78 78 78 78 78 78 78]
IDE0: NO MEDIA
IDE1: NO MEDIA
SD: MODE=MK4 FAST OPR=0x89 CNTR=0x4A TRDR=0x4B DEVICES=1
SD0: SDHC NAME=SS32G BLOCKS=0x03B72400 SIZE=30436MB
PRP: IO=0xA8 NOT PRESENT

Unit        Device      Type              Capacity/Mode
----------  ----------  ----------------  --------------------
Char 0      ASCI0:      RS-232            38400,8,N,1
Char 1      ASCI1:      RS-232            38400,8,N,1
Disk 0      MD0:        RAM Disk          256KB,LBA
Disk 1      MD1:        ROM Disk          384KB,LBA
Disk 2      IDE0:       Hard Disk         --
Disk 3      IDE1:       Hard Disk         --
Disk 4      SD0:        SD Card           30436MB,LBA


MARK IV (ide dbg) Boot Loader

Boot [H=Help]: 

---------------------------------------------
This is what the (failing) 32GB SanDisk Extreme PRO 160MB/s UDMA7 CF card looks like with TRACE set:-

RomWBW HBIOS v3.1.1-pre.177, 2022-07-02

MARK IV (ide dbg) Z8S180-N @ 18.432MHz IO=0x40
0 MEM W/S, 2 I/O W/S, INT MODE 2, Z180 MMU
512KB ROM, 512KB RAM
ROM VERIFY: 00 00 00 00 PASS

ASCI0: IO=0x40 ASCI W/BRG MODE=38400,8,N,1
ASCI1: IO=0x41 ASCI W/BRG MODE=38400,8,N,1
DSRTC: MODE=STD IO=0x8A Sun 2022-07-03 09:39:21 CHARGE=ON
CVDU: MODE=ECB EGA IO=0xE4 NOT PRESENT
VGA: IO=0xE2 NOT PRESENT
MD: UNITS=2 ROMDISK=384KB RAMDISK=256KB
FD: MODE=DIDE IO=0x2A NOT PRESENT
IDE: IO=0x80 MODE=MK4
IDE0: RESET FAKE HARD CONFIG
IDE0: SELUNIT
IDE0: PROBE E0 00 [00 00 00 00 00 00 00] [00 00 00 00 00 00 00] 00
IDE0: NO MEDIA [00 00 00 00 00 00 00]
IDE0: SELUNIT
IDE0: PROBE F0 78
IDE0: NO MEDIA [78 78 78 78 78 78 78]
IDE0: NO MEDIA
IDE1: NO MEDIA
SD: MODE=MK4 FAST OPR=0x89 CNTR=0x4A TRDR=0x4B DEVICES=1
SD0: SDHC NAME=SS32G BLOCKS=0x03B72400 SIZE=30436MB
PRP: IO=0xA8 NOT PRESENT

Unit        Device      Type              Capacity/Mode
----------  ----------  ----------------  --------------------
Char 0      ASCI0:      RS-232            38400,8,N,1
Char 1      ASCI1:      RS-232            38400,8,N,1
Disk 0      MD0:        RAM Disk          256KB,LBA
Disk 1      MD1:        ROM Disk          384KB,LBA
Disk 2      IDE0:       Hard Disk         --
Disk 3      IDE1:       Hard Disk         --
Disk 4      SD0:        SD Card           30436MB,LBA


MARK IV (ide dbg) Boot Loader

Boot [H=Help]:

**Reply by wwarthen** [reply]
Thanks Graeme,

Indeed this indicates a hardware incompatibility with the failing cards.  The first card is performing perfectly.  The second two are not responding at all. I have a couple cards that are also failing like yours.  These cards also fail under John's UNA BIOS, so there are multiple BIOSes that fail, so I'm pretty sure it is hardware.  I have tried various changes in the RomWBW BIOS and nothing seems to get these cards working.  These cards DO work fine when I plug them into a Dual IDE board in the Mark IV.  Again, points to a hardware interface issue on the CPU card's CF Interface.  Wish I had a solution, but I don't.  This interface is going to always prefer older cards.

This is the line from your working card that shows it is responding:

`IDE0: PROBE E0 50 [50 E0 00 00 01 01 01] [50 E0 00 00 01 01 01] 50 01 01 00 00`

These are the lines from your failing cards that show they do not respond at all:

`IDE0: PROBE E0 00 [00 00 00 00 00 00 00] [00 00 00 00 00 00 00] 00`

`IDE0: PROBE E0 00 [00 00 00 00 00 00 00] [00 00 00 00 00 00 00] 00`

Thanks,

Wayne

**Reply by sparkybrit** [reply]
Thanks for looking at this.
